src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,17 +357,65 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
+class Hook
+	constructor: (opts) ->
+		@runFn = opts.RUN
+		@rollbackFn = opts.ROLLBACK
+		@rollbackArgs = []
+		@executed = false
+
+	setArgument: (arg) ->
+		if @rollbackArgs.length < @rollbackFn.length
+			@rollbackArgs.push(arg)
+		else
+			throw new Error('Too many arguments supplied to rollback function')
+
+	run: (args...) ->
+		Promise.try =>
+			@runFn.apply(null, args)
+		.tap =>
+			if @rollbackArgs.length != @rollbackFn.length
+				throw new Error('registerRollback not supplied enough arguments')
+			@executed = true
+		# If we detect that some rollbacks are registered but the rollback function is not saturated, we add the missing number of arguments as null.
+		# The ROLLBACK function should take care of checking and performing the correct number of rollback actions.
+		.tapCatch =>
+			if not _.isEmpty(@rollbackArgs)
+				missingArgs = @rollbackFn.length - @rollbackArgs.length
+				for i in _.range(missingArgs)
+					@rollbackArgs.push(null)
+				@rollback()
+
+	rollback: ->
+		Promise.try =>
+			@rollbackFn.apply(null, @rollbackArgs)
+
 runHooks = (hookName, args) ->
 	Object.defineProperty args, 'api',
 		get: _.once ->
 			return api[args.request.vocabulary].clone(passthrough: _.pick(args, 'req', 'tx'))
 	hooks = args.request?.hooks?[hookName] || []
 	Promise.map hooks, (hook) ->
-		hook(args)
+		args.registerRollback = (args...) ->
```

Comment:
This seems like it would make more sense to just be:
```coffee
args.registerRollback = hook.registerRollback
```
and have the hook class handle it directly.  Alternatively it should be possible to have hooks be able to run `this.registerRollback()` and have them run in the context of the hook obj

Reply:
Yeah, I did it like this as initially registerRollback would need a reference to the current `request`. I like `@registerRollback()` in the hooks more than setting it on the args tbh.